src/main/java/org/spdx/utility/compare/TemplateRegexGenerator.java

```diff
@@ -0,0 +1,322 @@
+ * <code>getStartRegex(int wordLimit)</code> will return a regular expression to match the beginning of a license
+ * and <code>getEndRegex(int wordLimit)</code> will return a regular expression to match the end of a license
+ * 
+ * @author Gary O'Neall
+ *
+ */
+public class TemplateRegexGenerator implements ILicenseTemplateOutputHandler {
+	
+	static final Logger logger = LoggerFactory.getLogger(TemplateRegexGenerator.class);
+	
+	static final int WORD_LIMIT = 25; // number of words to search for at the beginning and end of the template
+	
+	static abstract class RegexElement {
+	}
+	
+	static class RegexList extends RegexElement {
+		private List<RegexElement> elements = new ArrayList<>();
+		
+		public void addElement(RegexElement element) {
+			elements.add(element);
+		}
+		
+		public List<RegexElement> getElements() {
+			return elements;
+		}
+		
+		@Override
+		public String toString() {
+			StringBuilder sb = new StringBuilder();
+			for (RegexElement element:elements) {
+				sb.append(element.toString());
+			}
+			return sb.toString();
+		}
+	}
+	
+	/**
+	 * Group with the optional modifier
+	 */
+	static class OptionalRegexGroup extends RegexList {
+		
+		@Override
+		public String toString() {
+			return "(" + super.toString() + ")?";
+		}
+	}
+	
+	static class RegexToken extends RegexElement {
+		private String token;
+		
+		public RegexToken(String token) {
+			this.token = token.trim();
+		}
+		
+		public String getToken() {
+			return token;
+		}
+		
+		@Override
+		public String toString() {
+			return Pattern.quote(
+					LicenseCompareHelper.NORMALIZE_TOKENS.getOrDefault(token.toLowerCase(), token)) + 
+					"\\s*";
+		}
+	}
+	
+	static class RegexPattern extends RegexElement {
+		private String pattern;
+		
+		public RegexPattern(String pattern) {
+			this.pattern = pattern;
+		}
+		
+		public String getPattern() {
+			return pattern;
+		}
+		
+		@Override
+		public String toString() {
+			return "(" + pattern + ")"; // We always treat the pattern as a group
+		}
+
+		/**
+		 * @param pattern pattern to set
+		 */
+		public void setPattern(String pattern) {
+			this.pattern = pattern;
+		}
+	}
+
+	private String template;
+	
+	/**
+	 * Top level regex
+	 */
+	private RegexList regexPatternList = new RegexList();
+	
+	
+	private int optionalNestLevel = 0;
+	
+	private List<OptionalRegexGroup> optionalGroups = new ArrayList<>();
+
+	/**
+	 * Generates regular expressions from a license or exception template
+	 * @throws SpdxCompareException 
+	 */
+	public TemplateRegexGenerator(String template) throws SpdxCompareException {
+		this.template = template;
+		parseTemplate();
+	}
+	
+	/**
+	 * Parses the template generating the regular expression
+	 * @throws SpdxCompareException 
+	 */
+	private void parseTemplate() throws SpdxCompareException {
+		try {
+			SpdxLicenseTemplateHelper.parseTemplate(LicenseCompareHelper.removeCommentChars(template), this);
```

Comment:
Perhaps the `removeCommentChars` method would make more sense to be located here, rather than over in `LicenseCompareHelper`?  Circular dependencies like this (LicenseCompareHelper -> TemplateRegexGenerator -> LicenseCompareHelper), while valid Java, are a bit of a red flag imho.

Reply:
Good point - there are actually a lot of circular dependencies between the `TemplateRegexGenerator` and `LicenseCompareHelper` in the code based on how it was refactored.  `removeCommentChars` is called by other methods in the `LicenseCompareHelper`.  Probably the best design would be to split out the common code into a separate static helper class, but that would introduce incompatibilities in the public API.

I'll see if I can come up with something cleaner and compatible.

Reply:
I did a bit of refactoring to remove all of the dependencies from the `LicenseCompareHelper` to the `TemplateRegexGenerator` (soon to be renamed `TemplateRegexMatcher`) to only two calls to a single method which reduces the cycles considerably.  I left the 2 calls in for compatibility reasons.